{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "## Load Libraries\n",
    "import os\n",
    "import bioframe\n",
    "\n",
    "##-----------------------------Quest Allocation------------------------------##\n",
    "\n",
    "allocationID = 'b1042'\n",
    "partitionName = 'genomics'\n",
    "email = 'lucascarter2025@u.northwestern.edu' #'Your_Email@u.northwestern.edu'\n",
    "\n",
    "##----------------------------Directory Variables----------------------------##\n",
    "\n",
    "## HiC files stored in /projects/b1042/BackmanLab/Emily/HiC\n",
    "\n",
    "# Local machine dir here the scripts and directory structer is initially generated\n",
    "output_path = '/Users/lucascarter/Documents/IBiS/Backman_Lab/projects/Emily/Lamins_HiC/HiC_Analysis_Dir/' # '/Users/Your_Name/Documents/Your_Working_Directory/'\n",
    "\n",
    "# Location of juicer directory on cluster\n",
    "juicer_dir = '/projects/b1042/BackmanLab/HiC2/opt/juicer' # '~/HiC/opt/juicer'\n",
    "\n",
    "# Location of non-Juicer post-Hi-C analysis python/R code on cluster\n",
    "analysis_code_dir = '/projects/b1042/BackmanLab/HiC2/opt/code_files/' # '~/HiC/code_files/'\n",
    "\n",
    "# Location of working juicer directory on cluster where FASTQs will be processed\n",
    "juicer_work_dir = '/projects/b1042/BackmanLab/HiC2/opt/juicer/work/' # '~/HiC/opt/juicer/work/'\n",
    "\n",
    "##----------------------------Experiment Variables----------------------------##\n",
    "\n",
    "# Directory name of the experiment you're analyzing\n",
    "experiment_name = 'Lamin_HiC' # 'Name_Of_Experiment'\n",
    "\n",
    "# A list of condition directories for each condition to be analyzed\n",
    "conds = ['untreated','24hrAuxin', 'withdraw'] # ['Cond_1','Cond_2', ect..]\n",
    "\n",
    "# A list of the number of replicates in each condition\n",
    "reps_per_cond = [6,6,6] # [1,1]\n",
    "\n",
    "# Name of genome to align to\n",
    "genome = \"hg19\" # 'genome_name'\n",
    "\n",
    "# Name of genome fa (if needed)\n",
    "fasta_name = \"GRCh19.primary_assembly.genome.fa\" # 'name_of_fasta'\n",
    "\n",
    "# Chromosomes to analyze contacts (depends on genome)\n",
    "chroms = {1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,'X'}\n",
    "\n",
    "# Name of txt file with restriction enzyme cuts of genome\n",
    "restriction_sites =  'hg19_DpnII.txt' # \"My_Restriction_Site_File.txt\"\n",
    "\n",
    "# Name of restriction enzyme\n",
    "restriction_enzyme = 'DpnII' #'restriction_enzyme'\n",
    "\n",
    "# Either \"Rep\" (analyze by replicate) or \"mega\" (pool all replicates together, analyze by condition)\n",
    "keyword = \"Rep\" # 'Rep_or_Mega'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "## Here, we create the directory structure for the HiC analysis ------------------------------------------##\n",
    "\n",
    "##### Input: output file path, experiment name, conditions analyzed, replicates per condition\n",
    "\n",
    "##### Output: Create correct directory/subdirectory structure\n",
    "\n",
    "##--------------------------------------------------------------------------------------------------------##\n",
    "\n",
    "def create_dir_struct(output_path,experiment_name,conds,reps_per_cond):\n",
    "\n",
    "    experiment_dir = output_path+experiment_name\n",
    "    if not os.path.isdir(experiment_dir):\n",
    "        os.mkdir(experiment_dir)\n",
    "\n",
    "    ## Run through each subdirectory and make sure it is there. If it isn't, mkdir\n",
    "    subdirnames = ['juicer_analysis','contact_data','contact_domains','compartment_analysis']\n",
    "\n",
    "    for subdirname in subdirnames:\n",
    "        subdir = experiment_dir+'/'+subdirname+'/'\n",
    "\n",
    "        if not os.path.isdir(subdir):\n",
    "            os.mkdir(subdir)\n",
    "\n",
    "## Here we generate the Bash scripts to process the fastq files using Juicer -----------------------------##\n",
    "\n",
    "##### Input: output file path, Juicer software and working directories, experiment name, conditions analyzed, replicates per condition, aligned genome, file containing restriction sites for specified genome\n",
    "\n",
    "##### Output: Make cluster files to perform juicer analysis on fastq files for each replicate in each condition\n",
    "\n",
    "##--------------------------------------------------------------------------------------------------------##\n",
    "\n",
    "def juicer_analysis_files(output_path,juicer_dir,juicer_work_dir,experiment_name,conds,reps_per_cond,genome,restriction_sites):\n",
    "    juicer_subdir = \"/juicer_analysis/\"\n",
    "\n",
    "    ## As before, check that directories exist. If not, make directories\n",
    "    for i,cond in enumerate(conds):\n",
    "        cond_dir = output_path+experiment_name+juicer_subdir+cond\n",
    "        if not os.path.isdir(cond_dir):\n",
    "            os.mkdir(cond_dir)\n",
    "\n",
    "        for rep in range(1,reps_per_cond[i]+1):\n",
    "            rep_dir = cond_dir+'/Rep'+str(rep)\n",
    "            if not os.path.isdir(rep_dir):\n",
    "                os.mkdir(rep_dir)\n",
    "\n",
    "            rep_dir_cluster = juicer_work_dir+experiment_name+juicer_subdir+cond+\"/Rep\"+str(rep)\n",
    "\n",
    "            if not os.path.isdir(rep_dir+'/fastq'):\n",
    "                os.mkdir(rep_dir+'/fastq')\n",
    "                \n",
    "            if not os.path.isdir(rep_dir+'/aligned'):\n",
    "                os.mkdir(rep_dir+'/aligned')\n",
    "\n",
    "            ## Open and write file for each rep called \"juicer_analysis.sh\"\n",
    "            f = open(rep_dir+'/juicer_analysis.sh','w')\n",
    "\n",
    "            f.write(\n",
    "\n",
    "\"#!/bin/bash\\n\\\n",
    "#SBATCH -A \"+allocationID+\"\\n\\\n",
    "#SBATCH -p \"+partitionName+\"\\n\\\n",
    "#SBATCH -N 1\\n\\\n",
    "#SBATCH --mem=100GB\\n\\\n",
    "#SBATCH --ntasks-per-node=16\\n\\\n",
    "#SBATCH --mail-user=\"+email+\"\\n\\\n",
    "#SBATCH --mail-type=BEGIN\\n\\\n",
    "#SBATCH --mail-type=END\\n\\\n",
    "#SBATCH -t 24:00:00\\n\\\n",
    "#SBATCH --job-name=\"+experiment_name+\"_\"+cond+\"_Rep\"+str(rep)+\"_juicer_analysis\\n\\n\\\n",
    "# Load necessary modules\\n\\\n",
    "module load bwa/0.7.17\\n\\n\\\n",
    "# Set your working directory\\n\\\n",
    "cd \"+rep_dir_cluster+\"\\n\\n\\\n",
    "# Perform Juicer Analysis on fastq files:\\n\"\n",
    "+juicer_dir+\"/scripts/juicer.sh -D \"+juicer_dir+\" -g \"+genome+\" -y \"+juicer_dir+\"/restriction_sites/\"+restriction_sites+\" -p \"+juicer_dir+\"/chrom_sizes/\"+genome+\".chrom.sizes\\n\\n\\\n",
    "# Perform cleanup on analysis files:\\n\"\n",
    "+juicer_dir+\"/scripts/common/cleanup.sh\\n\\n\\\n",
    "# Unzip merged_nodups file to perform mega analysis:\\n\\\n",
    "gunzip aligned/merged_nodups.txt.gz\")\n",
    "\n",
    "            f.close()\n",
    "\n",
    "## This function merges replicates processed in into a single matrix by condition -------------------------##\n",
    "\n",
    "##### Input: output file path, Juicer software and working directories, experiment name, conditions analyzed, replicates per condition, aligned genome, restriction enzyme\n",
    "\n",
    "##### Output: Make cluster files to merge Juicer analysis for replicates to create mega file for each condition\n",
    "\n",
    "##--------------------------------------------------------------------------------------------------------##\n",
    "\n",
    "def mega_files(output_path,juicer_dir,juicer_work_dir,experiment_name,conds,reps_per_cond,genome,restriction_enzyme):\n",
    "    juicer_subdir = \"/juicer_analysis/\"\n",
    "\n",
    "    ## As before, check that directories exist. If not, make directories\n",
    "    for i,cond in enumerate(conds):\n",
    "        if (reps_per_cond[i]>1):\n",
    "            cond_dir = output_path+experiment_name+juicer_subdir+cond\n",
    "            if not os.path.isdir(cond_dir):\n",
    "                os.mkdir(cond_dir)\n",
    "\n",
    "            cond_dir_cluster = juicer_work_dir+experiment_name+juicer_subdir+cond\n",
    "\n",
    "            for rep in range(1,reps_per_cond[i]+1):\n",
    "                rep_dir = cond_dir+'/Rep'+str(rep)\n",
    "                if not os.path.isdir(rep_dir):\n",
    "                    os.mkdir(rep_dir)\n",
    "\n",
    "                ## Open and write file for each rep called \"mega.sh\"\n",
    "                f = open(cond_dir+'/mega.sh','w')\n",
    "\n",
    "                f.write(\n",
    "\"#!/bin/bash\\n\\\n",
    "#SBATCH -A \"+allocationID+\"\\n\\\n",
    "#SBATCH -p \"+partitionName+\"\\n\\\n",
    "#SBATCH -N 1\\n\\\n",
    "#SBATCH --ntasks-per-node=16\\n\\\n",
    "#SBATCH --mail-user=\"+email+\"\\n\\\n",
    "#SBATCH --mail-type=BEGIN\\n\\\n",
    "#SBATCH --mail-type=END\\n\\\n",
    "#SBATCH -t 48:00:00\\n\\\n",
    "#SBATCH --job-name=\"+experiment_name+\"_\"+cond+\"_mega\\n\\n\\\n",
    "# Set your working directory\\n\\\n",
    "cd \"+cond_dir_cluster+\"\\n\\n\\\n",
    "# Merge hi-c analysis by replicate into mega analysis by condition:\\n\"\n",
    "+juicer_dir+\"/scripts/common/mega.sh -D \"+juicer_dir+\" -g \"+genome+\" -b \"+restriction_enzyme+\"\\n\\\n",
    "gzip \"+cond_dir_cluster+'/mega/aligned/merged')\n",
    "\n",
    "                f.close()\n",
    "\n",
    "## function to generate bash scripts that calculate map resolution for replicates and merged files -------##\n",
    "\n",
    "##### Input: output file path, Juicer software and Juicer working directory, experiment name, conditions analyzed, replicates per condition\n",
    "\n",
    "##### Output: Make cluster files to perform 3D-DNA analysis on merged_nodups.txt files from Juicer analysis for each replicate in each condition\n",
    "\n",
    "##--------------------------------------------------------------------------------------------------------##\n",
    "\n",
    "def calculate_map_resolution(output_path,juicer_dir,juicer_work_dir,experiment_name,conds,reps_per_cond):\n",
    "    juicer_subdir = \"/juicer_analysis/\"\n",
    "\n",
    "    ## As before, check that directories exist. If not, make directories\n",
    "    for i,cond in enumerate(conds):\n",
    "        cond_dir = output_path+experiment_name+juicer_subdir+cond\n",
    "        if not os.path.isdir(cond_dir):\n",
    "            os.mkdir(cond_dir)\n",
    "\n",
    "        if (keyword =='Rep'):\n",
    "\n",
    "            # Write get contacts file for each replicate\n",
    "            for rep in range(1,reps_per_cond[i]+1):\n",
    "                rep_dir = cond_dir+'/Rep'+str(rep)\n",
    "                if not os.path.isdir(rep_dir):\n",
    "                    os.mkdir(rep_dir)\n",
    "\n",
    "                juicer_rep_dir = juicer_work_dir+experiment_name+juicer_subdir+cond+\"/Rep\"+str(rep)\n",
    "\n",
    "                ## Open and write file for each rep called \"calculate_map_resolution.sh\"\n",
    "                f = open(rep_dir+'/calculate_map_resolution.sh','w')\n",
    "\n",
    "                ## This loop writes generates Bash scripts to calculate map resolution for replicates\n",
    "                f.write(\n",
    "\"#!/bin/bash\\n\\\n",
    "#SBATCH -A \"+allocationID+\"\\n\\\n",
    "#SBATCH -p \"+partitionName+\"\\n\\\n",
    "#SBATCH -N 1\\n\\\n",
    "#SBATCH --ntasks-per-node=16\\n\\\n",
    "#SBATCH --mail-user=\"+email+\"\\n\\\n",
    "#SBATCH --mail-type=BEGIN\\n\\\n",
    "#SBATCH --mail-type=END\\n\\\n",
    "#SBATCH -t 4:00:00\\n\\\n",
    "#SBATCH --job-name=\"+experiment_name+\"_\"+cond+\"_Rep\"+str(rep)+\"_calculate_map_resolution\\n\\n\\\n",
    "# Set your working directory\\n\\\n",
    "cd \"+juicer_rep_dir+\"/aligned\\n\\n\\\n",
    "# Calculate map resolution:\\n\"\n",
    "+juicer_dir+\"/misc/calculate_map_resolution.sh merged_nodups.txt 50bp_coverage.txt\\n\")\n",
    "\n",
    "                f.close()\n",
    "## End script generator for replicate map calculate_map_resolution\n",
    "\n",
    "        elif (keyword == \"mega\"):\n",
    "            # Write get contacts file for mega file\n",
    "            mega_dir = cond_dir+\"/mega\"\n",
    "            if not os.path.isdir(mega_dir):\n",
    "                os.mkdir(mega_dir)\n",
    "\n",
    "            juicer_mega_dir = juicer_work_dir+experiment_name+juicer_subdir+cond+\"/mega\"\n",
    "\n",
    "            f = open(mega_dir+'/calculate_map_resolution_mega.sh','w')\n",
    "\n",
    "## This loop writes generates Bash scripts to calculate map resolution for merged files\n",
    "\n",
    "            f.write(\n",
    "\"#!/bin/bash\\n\\\n",
    "#SBATCH -A \"+allocationID+\"\\n\\\n",
    "#SBATCH -p \"+partitionName+\"\\n\\\n",
    "#SBATCH -N 1\\n\\\n",
    "#SBATCH --ntasks-per-node=16\\n\\\n",
    "#SBATCH --mail-user=\"+email+\"\\n\\\n",
    "#SBATCH --mail-type=BEGIN\\n\\\n",
    "#SBATCH --mail-type=END\\n\\\n",
    "#SBATCH -t 4:00:00\\n\\\n",
    "#SBATCH --job-name=\"+experiment_name+\"_\"+cond+\"_mega_calculate_map_resolution\\n\\n\\\n",
    "# Set your working directory\\n\\\n",
    "cd \"+juicer_mega_dir+\"/aligned\\n\\n\\\n",
    "# Calculate map resolution:\\n\"\n",
    "+juicer_dir+\"/misc/calculate_map_resolution.sh merged_nodups.txt 50bp_coverage.txt\\n\")\n",
    "\n",
    "            f.close()\n",
    "## End script generator for mega(merged) map calculate_map_resolution\n",
    "\n",
    "        else:\n",
    "            print(\"Enter correct keyword\")\n",
    "\n",
    "## function to generate bash scripts that compress replicates --------------------------------------------##\n",
    "\n",
    "##### Input: output file path, Juicer working directory, experiment name, conditions analyzed, replicates per condition\n",
    "\n",
    "##### Output: Make cluster files to merge Juicer analysis for replicates to create mega file for each condition\n",
    "\n",
    "##--------------------------------------------------------------------------------------------------------##\n",
    "\n",
    "def compress_files(output_path,juicer_work_dir,experiment_name,conds,reps_per_cond):\n",
    "    juicer_subdir = \"/juicer_analysis/\"\n",
    "\n",
    "    ## As before, check that directories exist. If not, make directories\n",
    "    for i,cond in enumerate(conds):\n",
    "        cond_dir = output_path+experiment_name+juicer_subdir+cond\n",
    "        if not os.path.isdir(cond_dir):\n",
    "                os.mkdir(cond_dir)\n",
    "\n",
    "        cond_dir_cluster = juicer_work_dir+experiment_name+juicer_subdir+cond\n",
    "\n",
    "        for rep in range(1,reps_per_cond[i]+1):\n",
    "            rep_dir = cond_dir+'/Rep'+str(rep)\n",
    "            if not os.path.isdir(rep_dir):\n",
    "                os.mkdir(rep_dir)\n",
    "\n",
    "            f = open(cond_dir+'/compress_files.sh','w')\n",
    "\n",
    "            ## This loop writes generates Bash scripts to compress files\n",
    "            f.write(\n",
    "\"#!/bin/bash\\n\\\n",
    "#SBATCH -A \"+allocationID+\"\\n\\\n",
    "#SBATCH -p \"+partitionName+\"\\n\\\n",
    "#SBATCH -N 1\\n\\\n",
    "#SBATCH --ntasks-per-node=16\\n\\\n",
    "#SBATCH --mail-user=\"+email+\"\\n\\\n",
    "#SBATCH --mail-type=BEGIN\\n\\\n",
    "#SBATCH --mail-type=END\\n\\\n",
    "#SBATCH -t 48:00:00\\n\\\n",
    "#SBATCH --job-name=\"+experiment_name+\"_\"+cond+\"_compress_files\\n\\n\\\n",
    "# Set your working directory\\n\\\n",
    "cd \"+cond_dir_cluster+\"\\n\\n\\\n",
    "# Compress all stats files that were unzipped for mega analysis:\\n\\\n",
    "gzip **/aligned/*.txt\\n\")\n",
    "\n",
    "            f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "##------------- Call main method functions ----------------------------------------------------------------------##\n",
    "\n",
    "## Generate the directory structure where each of the initial analysis scripts will be stored\n",
    "create_dir_struct(output_path,experiment_name,conds,reps_per_cond)\n",
    "\n",
    "## Generate Juicer analysis files that process FASTQs into initial .HiC files\n",
    "juicer_analysis_files(output_path,juicer_dir,juicer_work_dir,experiment_name,conds,reps_per_cond,genome,restriction_sites)\n",
    "\n",
    "## Generate scripts to merge replicates in single .HiC files to increase resolution\n",
    "mega_files(output_path,juicer_dir,juicer_work_dir,experiment_name,conds,reps_per_cond,genome,restriction_enzyme)\n",
    "\n",
    "## Generate scripts to calculate resolution of HiC data\n",
    "calculate_map_resolution(output_path,juicer_dir,juicer_work_dir,experiment_name,conds,reps_per_cond)\n",
    "\n",
    "## Generate scripts to compress files for ease of storage and use\n",
    "compress_files(output_path,juicer_work_dir,experiment_name,conds,reps_per_cond)\n",
    "\n",
    "##---------------------------------------------------------------------------------------------------------------##\n",
    "\n",
    "\n",
    "##------------- Define Secondary Functions ----------------------------------------------------------------------##\n",
    "\n",
    "## function to generate bash scripts for calling domains with arrowhead -----------------------------------------##\n",
    "\n",
    "##### Input: output file path, Juicer software and working directories, experiment name, conditions analyzed, replicates per condition, keyword (either \"Rep\" or \"mega\"), matrix resolution and normalization method\n",
    "\n",
    "##### Output: Make cluster files to obtain contact information from Hi-C maps generated from Juicer Analysis\n",
    "\n",
    "##---------------------------------------------------------------------------------------------------------------##\n",
    "\n",
    "def get_contact_domains_arrowhead(output_path,juicer_dir,juicer_work_dir,experiment_name,conds,reps_per_cond,keyword,res,matrix_norm):\n",
    "    juicer_subdir = \"/juicer_analysis/\"\n",
    "    cont_domain_subdir = \"/contact_domains/arrowhead_domains/\"\n",
    "    if not os.path.isdir(output_path+experiment_name+cont_domain_subdir):\n",
    "            os.mkdir(output_path+experiment_name+cont_domain_subdir)\n",
    "\n",
    "    for i,cond in enumerate(conds):\n",
    "        cond_dir = output_path+experiment_name+cont_domain_subdir+cond\n",
    "\n",
    "        if not os.path.isdir(cond_dir):\n",
    "            os.mkdir(cond_dir)\n",
    "\n",
    "        if (keyword =='Rep'):\n",
    "            # Write get contacts file for each replicate\n",
    "            for rep in range(1,reps_per_cond[i]+1):\n",
    "                rep_dir = cond_dir+'/Rep'+str(rep)\n",
    "                if not os.path.isdir(rep_dir):\n",
    "                    os.mkdir(rep_dir)\n",
    "\n",
    "                juicer_rep_dir = juicer_work_dir+experiment_name+juicer_subdir+cond+\"/Rep\"+str(rep)\n",
    "                cont_domain_rep_dir = juicer_work_dir+experiment_name+cont_domain_subdir+cond+\"/Rep\"+str(rep)\n",
    "\n",
    "                f = open(rep_dir+'/contact_domains_'+matrix_norm+'_norm.sh','w')\n",
    "                f.write(\"#!/bin/bash\\n\\\n",
    "#SBATCH -A \"+allocationID+\"\\n\\\n",
    "#SBATCH -p \"+partitionName+\"\\n\\\n",
    "#SBATCH -N 1\\n\\\n",
    "#SBATCH --ntasks-per-node=16\\n\\\n",
    "#SBATCH --mail-user=\"+email+\"\\n\\\n",
    "#SBATCH --mail-type=BEGIN\\n\\\n",
    "#SBATCH --mail-type=END\\n\\\n",
    "#SBATCH -t 48:00:00\\n\\\n",
    "#SBATCH --job-name=\"+experiment_name+\"_\"+cond+\"_Rep\"+str(rep)+\"get_contact_domains\\n\\n\\\n",
    "# Set your working directory\\n\\\n",
    "cd \"+juicer_rep_dir+\"/aligned\\n\\n\\\n",
    "# Arrowhead annotation of contact domains:\\n\"\n",
    "+juicer_dir+\"/scripts/common/juicer_tools arrowhead --ignore_sparsity -r \"+str(res*1000)+\" -k \"+matrix_norm+\" rep\"+str(rep)+\"_inter_30.hic \"+cont_domain_rep_dir+\"/inter_30_contact_domains\\n\")\n",
    "                f.close()\n",
    "\n",
    "        elif (keyword == \"mega\"):\n",
    "            # Get contact domains for mega file\n",
    "            mega_dir = cond_dir+\"/mega\"\n",
    "            if not os.path.isdir(mega_dir):\n",
    "                os.mkdir(mega_dir)\n",
    "\n",
    "            juicer_mega_dir = juicer_work_dir+experiment_name+juicer_subdir+cond+\"/mega\"\n",
    "            cont_domain_mega_dir = juicer_work_dir+experiment_name+cont_domain_subdir+cond+\"/mega\"\n",
    "\n",
    "            f = open(mega_dir+'/contact_domains_'+matrix_norm+'_norm_mega.sh','w')\n",
    "            f.write(\"#!/bin/bash\\n\\\n",
    "#SBATCH -A \"+allocationID+\"\\n\\\n",
    "#SBATCH -p \"+partitionName+\"\\n\\\n",
    "#SBATCH -N 1\\n\\\n",
    "#SBATCH --ntasks-per-node=16\\n\\\n",
    "#SBATCH --mail-user=\"+email+\"\\n\\\n",
    "#SBATCH --mail-type=BEGIN\\n\\\n",
    "#SBATCH --mail-type=END\\n\\\n",
    "#SBATCH -t 48:00:00\\n\\\n",
    "#SBATCH --job-name=\"+experiment_name+\"_\"+cond+\"_mega_get_contact_domains\\n\\n\\\n",
    "# Set your working directory\\n\\\n",
    "cd \"+juicer_mega_dir+\"/aligned\\n\\n\\\n",
    "# Arrowhead annotation of contact domains:\\n\"\n",
    "+juicer_dir+\"/scripts/common/juicer_tools arrowhead --ignore_sparsity -r \"+str(res*1000)+\" -k \"+matrix_norm+\" Rep\"+str(rep)+\"_inter_30.hic \"+cont_domain_mega_dir+\"/inter_30_contact_domains\\n\")\n",
    "            f.close()\n",
    "\n",
    "        else:\n",
    "            print(\"Enter correct keyword\")\n",
    "\n",
    "## function to get contacts from .hic files  -----------------------------------------##\n",
    "\n",
    "##### Input: output file path, Juicer software and working directories, experiment name, conditions analyzed, replicates per condition, aligned genome, chromosomes, restriction enzyme, resolution of contacts (in kb), keyword (either \"Rep\" or \"mega\"), matrix normalization method \n",
    "\n",
    "##### Output: Make cluster files to obtain contact information from Hi-C maps generated from Juicer Analysis\n",
    "\n",
    "##---------------------------------------------------------------------------------------------------------------##\n",
    "\n",
    "def get_contacts_files(output_path,juicer_dir,juicer_work_dir,experiment_name,conds,reps_per_cond,genome,chroms,restriction_enzyme,keyword,res,matrix_norm):\n",
    "    juicer_subdir = \"/juicer_analysis/\"\n",
    "    contacts_subdir = \"/contact_data/\"\n",
    "    \n",
    "    chromosomes = \"(\"+' '.join([str(chrom) for chrom in chroms])+\")\"\n",
    "    for i,cond in enumerate(conds):\n",
    "        cond_dir = output_path+experiment_name+contacts_subdir+cond\n",
    "        if not os.path.isdir(cond_dir):\n",
    "            os.mkdir(cond_dir)\n",
    "\n",
    "        if (keyword =='Rep'):\n",
    "            # Write get contacts file for each replicate\n",
    "            for rep in range(1,reps_per_cond[i]+1):\n",
    "                rep_dir = cond_dir+'/Rep'+str(rep)\n",
    "                if not os.path.isdir(rep_dir):\n",
    "                    os.mkdir(rep_dir)\n",
    "                \n",
    "                contacts_dir = rep_dir+'/contacts'\n",
    "                if not os.path.isdir(contacts_dir):\n",
    "                    os.mkdir(contacts_dir)\n",
    "                        \n",
    "                juicer_rep_dir = juicer_work_dir+experiment_name+juicer_subdir+cond+\"/Rep\"+str(rep)\n",
    "                contacts_rep_dir = juicer_work_dir+experiment_name+contacts_subdir+cond+\"/Rep\"+str(rep)+\"/contacts\"\n",
    "\n",
    "                f = open(contacts_dir+'/contacts_'+matrix_norm+'_norm_'+str(res)+'kb.sh','w')\n",
    "                f.write(\"#!/bin/bash\\n\\\n",
    "#SBATCH -A \"+allocationID+\"\\n\\\n",
    "#SBATCH -p \"+partitionName+\"\\n\\\n",
    "#SBATCH -N 1\\n\\\n",
    "#SBATCH --ntasks-per-node=16\\n\\\n",
    "#SBATCH --mail-user=\"+email+\"\\n\\\n",
    "#SBATCH --mail-type=BEGIN\\n\\\n",
    "#SBATCH --mail-type=END\\n\\\n",
    "#SBATCH -t 4:00:00\\n\\\n",
    "#SBATCH --job-name=\"+experiment_name+\"_\"+cond+\"_Rep\"+str(rep)+\"_get_contacts\\n\\n\\\n",
    "# Set your working directory\\n\\\n",
    "cd \"+juicer_rep_dir+\"/aligned\\n\\n\\\n",
    "# Get contacts from Hi-C maps by chromosome:\\n\\\n",
    "chromosomes=\"+chromosomes+\"\\n\\\n",
    "for i in \\\"${chromosomes[@]}\\\"\\n\\\n",
    "do\\n\\\n",
    "    java -jar \"+juicer_dir+\"/scripts/common/juicer_tools.jar dump observed \"+matrix_norm+\" inter_30.hic $i $i BP \"+str(res*1000)+\" \"+contacts_rep_dir+\"/chr${i}-observed_\"+str(res)+\"Kb.txt\\n\\\n",
    "done\\n\")\n",
    "                f.close()\n",
    "        \n",
    "        elif (keyword == \"mega\"):\n",
    "            # Write get contacts file for mega file\n",
    "            mega_dir = cond_dir+\"/mega\"\n",
    "            if not os.path.isdir(mega_dir):\n",
    "                os.mkdir(mega_dir)\n",
    "            \n",
    "            contacts_dir = mega_dir+'/contacts'\n",
    "            if not os.path.isdir(contacts_dir):\n",
    "                os.mkdir(contacts_dir)\n",
    "            \n",
    "            juicer_mega_dir = juicer_work_dir+experiment_name+juicer_subdir+cond+\"/mega\"\n",
    "            contacts_mega_dir = juicer_work_dir+experiment_name+contacts_subdir+cond+\"/mega/contacts\"\n",
    "\n",
    "            f = open(contacts_dir+'/contacts_mega_'+matrix_norm+'_norm_'+str(res)+'kb.sh','w')\n",
    "            f.write(\"#!/bin/bash\\n\\\n",
    "#SBATCH -A \"+allocationID+\"\\n\\\n",
    "#SBATCH -p \"+partitionName+\"\\n\\\n",
    "#SBATCH -N 1\\n\\\n",
    "#SBATCH --ntasks-per-node=16\\n\\\n",
    "#SBATCH --mail-user=\"+email+\"\\n\\\n",
    "#SBATCH --mail-type=BEGIN\\n\\\n",
    "#SBATCH --mail-type=END\\n\\\n",
    "#SBATCH -t 4:00:00\\n\\\n",
    "#SBATCH --job-name=\"+experiment_name+\"_\"+cond+\"_mega_get_contacts\\n\\n\\\n",
    "# Set your working directory\\n\\\n",
    "cd \"+juicer_mega_dir+\"/aligned\\n\\n\\\n",
    "# Get contacts from Hi-C maps by chromosome:\\n\\\n",
    "chromosomes=\"+chromosomes+\"\\n\\\n",
    "for i in \\\"${chromosomes[@]}\\\"\\n\\\n",
    "do\\n\\\n",
    "    java -jar \"+juicer_dir+\"/scripts/common/juicer_tools.jar dump observed \"+matrix_norm+\" inter_30.hic $i $i BP \"+str(res*1000)+\" \"+contacts_mega_dir+\"/chr${i}-observed_\"+str(res)+\"Kb.txt\\n\\\n",
    "done\\n\")\n",
    "            f.close()\n",
    "        \n",
    "        else:\n",
    "            print(\"Enter correct keyword\")\n",
    "\n",
    "############### Experiment input #################################################################################\n",
    "res = 25 #Matrix resolution in kbp\n",
    "matrix_norm = 'VC' # matrix normalization method (OPTIONS: NONE/VC/VC_SQRT/KR)\n",
    "\n",
    "\n",
    "##------------- Call second method functions --------------------------------------------------------------------##\n",
    "\n",
    "get_contact_domains_arrowhead(output_path,juicer_dir,juicer_work_dir,experiment_name,conds,reps_per_cond,keyword,res,matrix_norm)\n",
    "get_contacts_files(output_path,juicer_dir,juicer_work_dir,experiment_name,conds,reps_per_cond,genome,chroms,restriction_enzyme,keyword,res,matrix_norm)\n",
    "\n",
    "##---------------------------------------------------------------------------------------------------------------##\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
